### PR TITLE
Optimization bnd pressure computation

### PR DESCRIPTION
**Pull request by svchb**
For simulations involving complex and/or large structures outside of continuous flow areas in 3D it is beneficial to switch the computation from being based on the boundary to be based on the fluid particles. 
In my simulations involving falling droplets, for example, this speeds up the simulation by 50 to 60%!

### CODE REVIEW

File: src/schemes/boundary/dummy_particles/dummy_particles.jl [hunk 1]
```diff
@@ -324,13 +324,32 @@ function compute_pressure!(boundary_model, ::AdamiPressureExtrapolation,
...
-        adami_pressure_extrapolation!(boundary_model, system, neighbor_system,
-                                      system_coords, neighbor_coords,
-                                      v_neighbor_system, nhs)
+        # This is an optimization for simulations with large and complex boundaries.
+        # Especially, in 3D simulations with large and/or complex structures outside
+        # of areas with permanent flow.
+        # Note: The version iterating neighbors first is not thread parallizable.
+        # The factor is based on the achievable speed-up of the thread parallizable version.
+        if nparticles(system) >
+           ceil(Int, 0.5 * Threads.nthreads()) * nparticles(neighbor_system)
```
Comment (efaulhaber) [inline src/schemes/boundary/dummy_particles/dummy_particles.jl:335]: Why the 0.5?

File: src/schemes/boundary/dummy_particles/dummy_particles.jl [hunk 2]
```diff
@@ -324,13 +324,32 @@ function compute_pressure!(boundary_model, ::AdamiPressureExtrapolation,
...
+            nhs = get_neighborhood_search(neighbor_system, system, semi)
+
+            adami_pressure_extrapolation_neighbor!(boundary_model, system, neighbor_system,
```
Comment (efaulhaber) [inline src/schemes/boundary/dummy_particles/dummy_particles.jl:338]:
```suggestion
            # Loop over fluid particles and then the neighboring boundary particles to extrapolate fluid pressure to the boundaries
            adami_pressure_extrapolation_neighbor!(boundary_model, system, neighbor_system,
```

File: src/schemes/boundary/dummy_particles/dummy_particles.jl [hunk 3]
```diff
@@ -324,13 +324,32 @@ function compute_pressure!(boundary_model, ::AdamiPressureExtrapolation,
...
+                                                   system_coords, neighbor_coords,
+                                                   v_neighbor_system, nhs)
+        else
+            nhs = get_neighborhood_search(system, neighbor_system, semi)
+
+            adami_pressure_extrapolation!(boundary_model, system, neighbor_system,
```
Comment (efaulhaber) [inline src/schemes/boundary/dummy_particles/dummy_particles.jl:344]:
```suggestion
            # Loop over boundary particles and then the neighboring fluid particles to extrapolate fluid pressure to the boundaries
            adami_pressure_extrapolation!(boundary_model, system, neighbor_system,
```

File: src/schemes/boundary/dummy_particles/dummy_particles.jl [hunk 4]
```diff
@@ -324,13 +324,32 @@ function compute_pressure!(boundary_model, ::AdamiPressureExtrapolation,
...
+                                          system_coords, neighbor_coords,
+                                          v_neighbor_system, nhs)
+        end
+        @simd for particle in eachparticle(system)
```
Comment (efaulhaber) [inline src/schemes/boundary/dummy_particles/dummy_particles.jl:348]:
```suggestion
        for particle in eachparticle(system)
```
There is absolutely no difference for me:
```
julia> @btime f($boundary_system, $(boundary_system.boundary_model.pressure))
  12.708 μs (0 allocations: 0 bytes)

julia> @btime f_inbounds($boundary_system, $(boundary_system.boundary_model.pressure))
  12.708 μs (0 allocations: 0 bytes)

julia> @btime f_simd($boundary_system, $(boundary_system.boundary_model.pressure))
  12.708 μs (0 allocations: 0 bytes)
```
Did you see a speedup on your machine?

File: src/schemes/boundary/dummy_particles/dummy_particles.jl [hunk 5]
```diff
@@ -377,28 +425,9 @@ end
...
-
-        compute_smoothed_velocity!(cache, viscosity, neighbor_system, v_neighbor_system,
-                                   kernel_weight, particle, neighbor)
-    end
-
-    for particle in eachparticle(system)
-        # Limit pressure to be non-negative to avoid attractive forces between fluid and
-        # boundary particles at free surfaces (sticking artifacts).
-        pressure[particle] = max(pressure[particle], 0.0)
+        adami_pressure_inner!(boundary_model, system, neighbor_system::FluidSystem,
```
Comment (efaulhaber) [inline src/schemes/boundary/dummy_particles/dummy_particles.jl:428]:
```suggestion
        adami_pressure_inner!(boundary_model, system, neighbor_system,
```

File: src/schemes/boundary/dummy_particles/dummy_particles.jl [hunk 6]
```diff
@@ -365,6 +384,35 @@ function compute_pressure!(boundary_model, ::Union{PressureMirroring, PressureZe
...
+                                                        neighborhood_search)
+    (; pressure, cache, viscosity) = boundary_model
+
+    for_particle_neighbor(neighbor_system, system,
+                          neighbor_coords, system_coords,
+                          neighborhood_search;
+                          particles=eachparticle(neighbor_system),
+                          parallel=false) do neighbor, particle,
+                                             pos_diff, distance
+        # since neighbor and particle are switched
```
Comment (efaulhaber) [inline src/schemes/boundary/dummy_particles/dummy_particles.jl:400]:
```suggestion
        # Since neighbor and particle are switched
```
#133

**Review by efaulhaber — CHANGES_REQUESTED** [review]

File: src/schemes/boundary/dummy_particles/dummy_particles.jl (hunk 1 above)
Comment (svchb) [reply src/schemes/boundary/dummy_particles/dummy_particles.jl:335]: There is a typo though **parallelizable**

File: src/schemes/boundary/dummy_particles/dummy_particles.jl (hunk 4 above)
Comment (svchb) [reply src/schemes/boundary/dummy_particles/dummy_particles.jl:348]: I did not try this.

**Review by efaulhaber — APPROVED** [review]

**Pull request merged**